hashes/zkevm/src/keccak/mod.rs

```diff
@@ -568,6 +553,111 @@ impl<F: Field> KeccakCircuitConfig<F> {
+            meta.query_fixed(col, Rotation::cur())
+        };
+        let q_in_round = |col: Column<Fixed>, meta: &mut VirtualCells<'_, F>| {
+            (0..num_rows_per_round as i32)
+                .map(|i| meta.query_fixed(col, Rotation(-i)))
+                .fold(0.expr(), |acc, elem| acc + elem)
+        };
+
+        /*
+        eg：
+            data:
+                get_num_rows_per_round: 18
+                input: "12345678abc"
+            table:
+                Note[1]: be careful: is_paddings is not column here! It is [Cell; 8] and it will be constrained later.
+        offset word_value bytes_left  is_paddings q_enable q_padding_last
+        18     0x87654321    11          0         1        0 // 1st round begin
+        19        0          10          0         0        0
+        20        0          9           0         0        0
+        21        0          8           0         0        0
+        22        0          7           0         0        0
+        23        0          6           0         0        0
+        24        0          5           0         0        0
+        25        0          4           0         0        0
+        26        0          4           NA        0        0
+        ...
+        35        0          4           NA        0        0  // 1st round end
+        36      0xcba        3           0         1        1  // 2nd round begin
+        37        0          2           0         0        0
+        38        0          1           0         0        0
+        39        0          0           1         0        0
+        40        0          0           1         0        0
+        41        0          0           1         0        0
+        42        0          0           1         0        0
+        43        0          0           1         0        0
+        */
+
+        meta.create_gate("word_value", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            cb.require_equal(
+                "word value",
+                absorb_data.expr(),
+                meta.query_advice(keccak_table.word_value, Rotation::cur()),
+            );
+            cb.gate(q(q_enable, meta) * not::expr(q(q_padding, meta)))
+        });
+        meta.create_gate("bytes_left", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            // bytes_left[i] == 0 in the first round
+            cb.condition(q_in_round(q_first, meta), |cb| {
+                cb.require_zero(
+                    "bytes_left needs to be zero on the first row",
+                    meta.query_advice(keccak_table.bytes_left, Rotation::cur()),
+                );
+            });
+            // is_paddings only be true when
+            cb.condition(q(q_padding, meta), |cb| {
+                for i in 0..num_rows_per_round {
+                    let bytes_left = meta.query_advice(keccak_table.bytes_left, Rotation(i as i32));
```

Comment:
As discussed in `tests.rs`, this is still recording `bytes_left` for each input byte, which means `rows_per_round >= 8` assumption is necessary. 

I think it is possible, without increasing custom gate `MAX_DEGREE` to just update `bytes_left` once per round by summing over all of `is_paddings` in this custom gate?

Reply:
This part has been rewritten.